lib/rust/ensogl/core/src/display/render/passes/cache_shapes.rs

```diff
@@ -0,0 +1,314 @@
+                for shape in ready_to_render {
+                    with_context(|t| {
+                        t.set_camera(&self.layer.camera());
+                        t.update();
+                    });
+                    shape.sprite().symbol.render();
+                }
+                // Restore screen viewport.
+                instance.context.viewport(0, 0, instance.width, instance.height);
+                instance.context.bind_framebuffer(*Context::FRAMEBUFFER, None);
+            } else {
+                error!("Impossible happened: The CacheShapesPass was run without initialized framebuffer.")
+            }
+        }
+    }
+}
+
+
+
+// ===================================
+// === Arranging Shapes on Texture ===
+// ===================================
+
+/// For shape of given size, find the unoccupied space in the square texture.
+fn find_free_place(
+    shape_size: Vector2<i32>,
+    tex_size: i32,
+    placed_so_far: &[BoundingBox],
+) -> Option<BoundingBox> {
+    let right_bounds = placed_so_far.iter().map(|bbox| bbox.right().ceil() as i32);
+    let allowed_right_bounds = right_bounds.filter(|col| col + shape_size.x <= tex_size);
+    let top_bounds = placed_so_far.iter().map(|bbox| bbox.top().ceil() as i32);
+    let allowed_top_bounds = top_bounds.filter(|row| row + shape_size.y <= tex_size);
+    let candidate_rows = iter::once(0).chain(allowed_top_bounds);
+    let candidate_cols = iter::once(0).chain(allowed_right_bounds);
+    let candidate_positions = iproduct!(candidate_rows, candidate_cols);
+    let mut candidate_bboxes = candidate_positions.map(|(y, x)| {
+        BoundingBox::from_position_and_size(
+            Vector2(x as f32, y as f32),
+            shape_size.map(|i| i as f32),
+        )
+    });
+    candidate_bboxes.find(|bbox| {
+        let is_collision = placed_so_far.iter().any(|placed| placed.interior_intersects(bbox));
+        !is_collision
+    })
+}
+
+#[derive(Derivative)]
+#[derivative(Debug)]
+struct ArrangedShapes {
+    texture_width:  i32,
+    texture_height: i32,
+    #[derivative(Debug = "ignore")]
+    shapes:         Vec<Box<dyn AnyShapeView>>,
+}
+
+/// Arrange Cached Shapes on texture.
+///
+/// Algorithm of placing each shape on the texture is a simple heuristic, where we sort shapes from
+/// the biggest to the smallest, and take the shapes one by one by that order and try to fit them
+/// by scanning possible left-bottom corner positions, starting from left-bottom corner of the
+/// texture and going line-by-line. According to
+/// https://gamedev.stackexchange.com/questions/2829/texture-packing-algorithm this is actually one
+/// of the best packing algorithm in terms of texture space.
+///
+/// This is a simple implementation with time complexity O(n m s) where n and m is the number of
+/// rows and columns in the resulting texture, and s is number of placed shapes.
+fn arrange_shapes_on_texture<'a>(
```

Comment:
> you are placing them from left to right and from bottom to top, right? 

This is what I mean by "starting from left-bottom corner of the texture and going line-by-line" Perhaps I should write "row by row"? It's more accurate than "placing them from left to right and from bottom to top" because once filled the bottom row, the next shape may be placed on the right (if the rightmost shape placed so far is smaller than the rest).